### PR TITLE
Fix bug publish screenshot

### PR DESCRIPTION
**Pull request by languy**
[Preview this branch](https://cosmos-explorer-preview.azurewebsites.net/pull/762?feature.someFeatureFlagYouMightNeed=true)

The main change in this PR fixes the snapshot functionality in the Publish pane-related components. Because the code cell outputs are now rendered in their own iframes for security reasons, a single snapshot of the notebook is no longer possible: each cell output takes its own snapshot and the snapshots are collated on the main notebook snapshot.
- Move the snapshot functionality to notebook components: this removes the reference of the notebook DOM node that we must pass to the Publish pane via explorer.
- Add slice in the state and actions in notebook redux for notebook snapshot requests and result
- Add post robot message to take snapshots and receive results
- Add logic in `NotebookRenderer` to wait for all output snapshots done before taking the main one collating.
- Use `zustand` to share snapshot between Redux world and React world. This solves the issue of keeping the `PanelContainer` component generic, while being able to update its children (`PublishPanel` component) with the new snapshot.

Additional changes:
- Add `local()` in `@font-face` to check if font is already installed before downloading the font (must be done for Safari, but not Edge/Chrome)
- Add "Export output to image" menu item in notebook cell, since each cell output can take its own snapshot (which can be downloaded)
image

### CODE REVIEW

File: src/Explorer/Notebook/NotebookUtil.ts [hunk 1]
```diff
@@ -141,23 +144,175 @@ export class NotebookUtil {
...
+      }
+    });
+  };
+
+  public static takeScreenshotDomToImage = (
+    target: HTMLElement,
+    aspectRatio: number,
+    subSnapshots: SnapshotFragment[],
+    downloadFilename?: string
+  ): Promise<{ imageSrc: string | undefined }> => {
```
Comment (southpolesteve) [inline src/Explorer/Notebook/NotebookUtil.ts:246]:
```suggestion
  ): Promise<{ imageSrc?: string }> => {
```

File: src/Explorer/Panes/PublishNotebookPane/PublishNotebookPane.tsx [hunk 2]
```diff
@@ -74,6 +79,14 @@ export const PublishNotebookPane: FunctionComponent<PublishNotebookPaneAProps> =
     setContent(newContent);
   }, []);
 
+  useEffect(() => {
+    setImageSrc(notebookSnapshot);
+  }, [notebookSnapshot]);
+
+  useEffect(() => {
+    setFormError(notebookSnapshotError);
+  }, [notebookSnapshotError]);
```
Comment (southpolesteve) [inline src/Explorer/Panes/PublishNotebookPane/PublishNotebookPane.tsx:88]: This feels strange to me to have to set these inside a useEffect. But I am not exactly sure what to do differently. Maybe there is a larger state issue to fix that is out of scope for this PR :/

File: src/hooks/useNotebookSnapshotStore.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+export interface NotebookSnapshotHooks {
+  snapshot: string;
+  error: string;
+  setSnapshot: (imageSrc: string) => void;
+  setError: (error: string) => void;
+}
+
+export const useNotebookSnapshotStore: UseStore<NotebookSnapshotHooks> = create((set) => ({
+  snapshot: undefined as string,
+  error: undefined as string,
```
[the remark this answers is not in the document]
Comment (languy) [reply src/hooks/useNotebookSnapshotStore.ts:12]: My intent was to require these fields to exist, but allow the values to be `undefined`. For example no error means the error is undefined, not that there's no error field. I know that for testing with `if` it doesn't matter, it seems semantically different. Is that the right way to think about it?

File: src/hooks/useSidePanel.ts [hunk 4]
```diff
@@ -4,25 +4,31 @@ export interface SidePanelHooks {
...
-  openSidePanel: (headerText: string, panelContent: JSX.Element) => void;
+  openSidePanel: (headerText: string, panelContent: JSX.Element, onClose?: () => void) => void;
   closeSidePanel: () => void;
 }
 
 export const useSidePanel = (): SidePanelHooks => {
   const [isPanelOpen, setIsPanelOpen] = useState<boolean>(false);
   const [panelContent, setPanelContent] = useState<JSX.Element>();
   const [headerText, setHeaderText] = useState<string>();
+  const [onCloseCallback, setOnCloseCallback] = useState<{ callback: () => void }>();
```
[the remark this answers is not in the document]
Comment (languy) [reply src/hooks/useSidePanel.ts:15]: I tried this originally and it didn't work with a warning ("Warning: Cannot update a component (`ConnectFunction`) while rendering a different component (`App`). To locate the bad setState() call inside `App`, follow the stack trace as described in https://fb.me/setstate-in-render" with the stack trace pointing to that line).

It turns out, plugging a function as an argument to `useState` means lazy initializing. So another option around this is to plug a [function that returns our function to store in the state](https://medium.com/swlh/how-to-store-a-function-with-the-usestate-hook-in-react-8a88dd4eede1). But calling `setOnCloseCallback` must also by done with a function that returns `onClose` (it works, but not necessarily by design). I find wrapping the function in an object less convoluted as a function returning a function and also seems safer for the setter.

File: src/hooks/useNotebookSnapshotStore.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import create, { UseStore } from "zustand";
```
[the remark this answers is not in the document]
Comment (languy) [reply src/hooks/useNotebookSnapshotStore.ts:1]: Absolutely.

File: src/Explorer/Notebook/NotebookUtil.ts [hunk 6]
```diff
@@ -141,23 +144,175 @@ export class NotebookUtil {
...
+      return accumulator;
+    }, []);
+  }
+
+  public static takeScreenshotHtml2Canvas = (
+    target: HTMLElement,
+    aspectRatio: number,
+    subSnapshots: SnapshotFragment[],
+    downloadFilename?: string
+  ): Promise<{ imageSrc: string | undefined }> => {
```
[the remark this answers is not in the document]
Comment (languy) [reply src/Explorer/Notebook/NotebookUtil.ts]: Similar reason here: I would like the promise to always return `imageSrc` with `undefined` as a possible valid value (empty output = blank image), because it will be stored as such in the redux store. I think it's semantically different from saying it might return `imageSrc` or not. If the function fails (no `imageSrc` can be provided), the `reject` path should be called.

File: src/Explorer/Panes/PublishNotebookPane/PublishNotebookPane.tsx (hunk 2 above)
Comment (languy) [reply src/Explorer/Panes/PublishNotebookPane/PublishNotebookPane.tsx:88]: I admit that was the easiest way to transcribe the current behavior ("if one of these values change, update the corresponding state"), but could be overkill.
Conceptually, `imageSrc` and these zustand store state variables are both state variables and the true react way is probably to not update a state variable with another variable, but instead somehow compose them together in an expression at the time of use. Something like:
` const response = await junoClient.publishNotebook( .... , notebookSnapshot ?? imageSrc, ....)`

But yes, I'd prefer getting this in and iterate :).

File: src/Explorer/Notebook/NotebookComponent/types.ts [hunk 7]
```diff
@@ -1,15 +1,31 @@
...
+  image: HTMLImageElement;
+  boundingClientRect: DOMRect;
+  requestId: string;
+}
+
+export interface SnapshotRequest {
+  requestId: string;
+  aspectRatio: number;
+  type: "notebook" | "celloutput";
+  cellId: string; // if type is "celloutput"
```
[the remark this answers is not in the document]
Comment (languy) [reply src/Explorer/Notebook/NotebookComponent/types.ts:16]: Yes, makes sense.

File: src/hooks/useNotebookSnapshotStore.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (languy) [reply src/hooks/useNotebookSnapshotStore.ts]: After discussion with Steve, I'm changing all these to optional.

File: src/Explorer/Notebook/NotebookUtil.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (languy) [reply src/Explorer/Notebook/NotebookUtil.ts]: Changing to optional instead of `string | undefined`.

**Review by southpolesteve — APPROVED** [review]

**Review by gitanuj — CHANGES_REQUESTED** [review]
Export output to image is neat 💯

I noticed that if I select "Take screenshot" option when trying to publish the notebook nothing shows up in the preview and my browser tab starts hanging. Opening the Dev Console I see 1000s of network requests firing continuously (initiated by the screenshot logic). Looks like there's some issue with screenshot-ing the notebook?

File: src/Explorer/Notebook/NotebookRenderer/outputs/SandboxOutputs.tsx [hunk 8]
```diff
@@ -24,27 +26,47 @@ interface StateProps {
...
-      <IframeResizer
-        checkOrigin={false}
-        loading="lazy"
-        heightCalculationMethod="taggedElement"
-        onLoad={(event) => this.handleFrameLoad(event)}
-        src="./cellOutputViewer.html"
-        style={{ height: "1px", width: "1px", minWidth: "100%", border: "none" }}
-        sandbox="allow-downloads allow-popups allow-forms allow-pointer-lock allow-scripts allow-popups-to-escape-sandbox"
-      />
+    return this.props.outputs && this.props.outputs.size > 0 ? (
```
[the remark this answers is not in the document]
Comment (languy) [reply src/Explorer/Notebook/NotebookRenderer/outputs/SandboxOutputs.tsx:56]: The suggestion is valid in the case `this.props.outputs` is `undefined`, because `undefined > 0` is `true`, but I prefer keeping the check separate so we only do the `size > 0` comparison if `size` is defined (ie a valid number).

**Review by gitanuj — APPROVED** [review]
:shipit:
